src/Microsoft.Identity.Web/TokenCacheProviders/TokenCacheExtensions.cs

```diff
@@ -75,15 +81,21 @@ public static class TokenCacheExtensions
                 throw new ArgumentNullException(nameof(initializeCaches));
             }
 
-            IHostBuilder hostBuilder = Host.CreateDefaultBuilder()
-                .ConfigureLogging(logger => { })
-                .ConfigureServices(services =>
-                 {
-                     initializeCaches(services);
-                     services.AddDataProtection();
-                 });
+            // Maintain a dictionary of service providers per `initializeCaches` delegate.
+            if (!s_serviceProviderFromAction.TryGetValue(initializeCaches.Method, out IServiceProvider? serviceProvider))
+            {
+                IHostBuilder hostBuilder = Host.CreateDefaultBuilder()
+                    .ConfigureLogging(logger => { })
+                    .ConfigureServices(services =>
+                    {
+                        initializeCaches(services);
+                        services.AddDataProtection();
+                    });
+
+                serviceProvider = hostBuilder.Build().Services;
+                s_serviceProviderFromAction.Add(initializeCaches.Method, serviceProvider);
```

Comment:
Use `s_serviceProviderFromAction[initializeCaches.Method] = serviceProvider`. Add will throw an exception if you add the same key twice.

Reply:
fixed.